storage/src/main/java/fk/prof/storage/S3AsyncStorage.java

```diff
@@ -113,22 +81,32 @@ private void initialize() {
         S3ObjectPath objectPath = new S3ObjectPath(prefixPath);
         return CompletableFuture.supplyAsync(() -> {
             try {
+                long startTime = System.currentTimeMillis();
                 Set<String> allObjects = new HashSet<>();
-                ObjectListing objects;
-                if (recursive) {
-                    objects = client.listObjects(new ListObjectsRequest().withBucketName(objectPath.bucket).withPrefix(objectPath.fileName));
-                } else {
-                    objects = client.listObjects(new ListObjectsRequest().withBucketName(objectPath.bucket).withDelimiter(DELIMITER).withPrefix(objectPath.fileName));
-                }
-
+                ObjectListing objects = null;
+                ListObjectsRequest request = new ListObjectsRequest().withBucketName(objectPath.bucket).withPrefix(objectPath.fileName);
                 do {
+                    if(!recursive) {
+                        request.withDelimiter(DELIMITER);
+                    }
+                    if(objects != null && objects.isTruncated()) {
```

Comment:
Is objects.isTruncated() required here? Should always be true when objects!=null because checked in while condition below